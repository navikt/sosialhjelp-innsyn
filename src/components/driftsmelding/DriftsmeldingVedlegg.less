.driftsmelding-vedlegg-wrapper {
  display: flex;
  align-items: center;
  margin-top: 2rem;
  margin-bottom: 1rem;
}

.driftsmelding-vedlegg-symbol-wrapper {
  margin-right: 8px;
  height: 24px;
}

.driftsmelding-vedlegg-text-wrapper {
  font-weight: 600;
  font-family: 'Source Sans Pro', serif;
}