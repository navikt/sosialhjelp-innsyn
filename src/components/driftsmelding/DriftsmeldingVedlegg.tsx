import * as React from "react";
import {useTranslation} from "react-i18next";
import {isFileUploadAllowed} from "./DriftsmeldingUtilities";
import {Alert} from "@navikt/ds-react";
import styled from "styled-components/macro";
import useKommune from "../../hooks/useKommune";

const Bold = styled.span`
    font-weight: bold;
`;

const DriftsmeldingVedlegg = () => {
    const {kommune, isLoading} = useKommune();
    const {t} = useTranslation();
    const kanLasteOppVedlegg: boolean = isFileUploadAllowed(kommune);

    if (!kanLasteOppVedlegg && !isLoading) {
        return (
            <Alert variant="error" size="medium" inline>
                <Bold>{t("driftsmelding.kanIkkeSendeVedlegg")}</Bold>
            </Alert>
        );
    }
    return null;
};

export default DriftsmeldingVedlegg;
