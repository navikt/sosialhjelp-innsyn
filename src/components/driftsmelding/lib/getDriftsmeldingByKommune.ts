import { KommuneResponse } from "../../../generated/model";

const mldInnsynOgEttersendelseDeaktivert = "driftsmelding.innsynOgEttersendelseDeaktivert" as const;
const mldEttersendelseDeaktivert = "driftsmelding.ettersendelseDeaktivert" as const;
const mldInnsynDeaktivert = "driftsmelding.innsynDeaktivert" as const;

type KommuneDriftsmeldingError =
    | typeof mldInnsynOgEttersendelseDeaktivert
    | typeof mldEttersendelseDeaktivert
    | typeof mldInnsynDeaktivert;

// Man skulle trodd at erInnsynDeaktivert skulle telle her, men enhetstestene tester faktisk at denne skal returnere false selv om erInnsynDeaktivert er true.
const innsynDeaktivert = ({ erInnsynMidlertidigDeaktivert }: KommuneResponse) => erInnsynMidlertidigDeaktivert;

const ettersendelseDeaktivert = ({
    erInnsendingEttersendelseDeaktivert,
    erInnsendingEttersendelseMidlertidigDeaktivert,
}: KommuneResponse) => erInnsendingEttersendelseMidlertidigDeaktivert || erInnsendingEttersendelseDeaktivert;

type KommuneDriftstatus = `ettersendelseNede: ${boolean}, innsynNede: ${boolean}`;

const DRIFTSMELDINGER: Record<KommuneDriftstatus, KommuneDriftsmeldingError | undefined> = {
    "ettersendelseNede: true, innsynNede: true": mldInnsynOgEttersendelseDeaktivert,
    "ettersendelseNede: true, innsynNede: false": mldEttersendelseDeaktivert,
    "ettersendelseNede: false, innsynNede: true": mldInnsynDeaktivert,
    "ettersendelseNede: false, innsynNede: false": undefined,
} as const;

const getDriftsstatus = (kommuneResponse: KommuneResponse): KommuneDriftstatus =>
    `ettersendelseNede: ${ettersendelseDeaktivert(kommuneResponse)}, innsynNede: ${innsynDeaktivert(kommuneResponse)}`;

export const getDriftsmeldingByKommune = (
    kommuneResponse: KommuneResponse | undefined
): KommuneDriftsmeldingError | undefined =>
    kommuneResponse ? DRIFTSMELDINGER[getDriftsstatus(kommuneResponse)] : undefined;
