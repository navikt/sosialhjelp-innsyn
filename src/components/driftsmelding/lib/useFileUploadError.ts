import { useHentSoknadsStatus } from "@generated/soknads-status-controller/soknads-status-controller";
import { useHentKommuneInfo } from "@generated/kommune-controller/kommune-controller";
import { KommuneResponse } from "@generated/model";

import useFiksDigisosId from "../../../hooks/useFiksDigisosId";

const ettersendelseDeaktivert = ({
    erInnsendingEttersendelseDeaktivert,
    erInnsendingEttersendelseMidlertidigDeaktivert,
}: KommuneResponse) => erInnsendingEttersendelseMidlertidigDeaktivert || erInnsendingEttersendelseDeaktivert;

/**
 * Utleder feilstatus for filopplasting, basert på kommunal driftstatus
 * fra FIKS Digisos.
 *
 * @returns
 *    "driftsmelding.kanIkkeSendeVedlegg"
 *      dersom fiks digisos kommuneinfo sier at ettersendelse er deaktivert.
 *    undefined
 *      mens vi laster kommune- og søknadsinfo
 *    null
 *      om alt er OK.
 */
export const useFileUploadError = () => {
    const fiksDigisosId = useFiksDigisosId();
    const { data: kommuneInfo } = useHentKommuneInfo(fiksDigisosId);
    const { data: soknadStatus } = useHentSoknadsStatus(fiksDigisosId);

    if (!kommuneInfo || !soknadStatus) return undefined;
    if (ettersendelseDeaktivert(kommuneInfo)) return "driftsmelding.kanIkkeSendeVedlegg";
    return null;
};
