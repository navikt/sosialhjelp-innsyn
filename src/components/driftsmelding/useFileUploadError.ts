import { KommuneResponse } from "../../generated/model";
import { useHentSoknadsStatus } from "../../generated/soknads-status-controller/soknads-status-controller";

export const ettersendelseErDeaktivert = (kommuneInfo: KommuneResponse | undefined) => {
    return (
        !kommuneInfo ||
        kommuneInfo.erInnsendingEttersendelseMidlertidigDeaktivert ||
        kommuneInfo.erInnsendingEttersendelseDeaktivert
    );
};

export const useFileUploadError = (kommuneInfo: KommuneResponse | undefined, fiksDigisosId: string) => {
    const { data } = useHentSoknadsStatus(fiksDigisosId);
    if (!data) return undefined;
    else if (data.isBroken) return "driftsmelding.vedlegg.vedleggMangler";
    else if (ettersendelseErDeaktivert(kommuneInfo)) return "driftsmelding.kanIkkeSendeVedlegg";
    return null;
};
