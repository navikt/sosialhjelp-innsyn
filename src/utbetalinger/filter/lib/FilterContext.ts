import { createContext } from "react";

type Maybe<T> = T | null;

export type MottakerFilter = "minKonto" | "annenMottaker";

export type FilterCriteria = {
    mottaker?: Maybe<MottakerFilter>;
    fraDato?: Maybe<Date>;
    tilDato?: Maybe<Date>;
};

export type FilterKey = keyof FilterCriteria;

type FilterContextType = {
    filters: FilterCriteria | null;
    setFilter: (nyttFilter: FilterCriteria) => void;
    clearFilters: () => void;
};

export const FilterContext = createContext<FilterContextType | undefined>(undefined);
