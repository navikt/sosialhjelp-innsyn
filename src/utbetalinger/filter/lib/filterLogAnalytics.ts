import { AmplitudeFiltervalgEvent, logAmplitudeEventTyped } from "../../../utils/amplitude";

import { FilterKey, FilterCriteria } from "./FilterContext";

const getFilterUpdates = (action: FilterCriteria): AmplitudeFiltervalgEvent[] =>
    Object.keys(action)
        .filter((key) => action[key as FilterKey] !== undefined)
        .map((key) => ({
            eventName: "filtervalg",
            eventData: {
                kategori: key as FilterKey,
                filternavn: action[key as FilterKey],
            },
        }));

export const filterLogAnalytics = (action: FilterCriteria) =>
    getFilterUpdates(action).map((e) => logAmplitudeEventTyped(e));
