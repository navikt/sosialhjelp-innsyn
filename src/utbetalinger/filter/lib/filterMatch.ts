import { isAfter, isBefore } from "date-fns";

import { ManedUtbetaling } from "../../../generated/model";

import { FilterPredicate } from "./FilterContext";

const stringToDateWithoutTimezone = (datoString: string | undefined) =>
    !datoString ? undefined : new Date(new Date(datoString).toISOString().slice(0, -1));

export const filterMatch = (
    {
        annenMottaker,
        forfallsdato,
        utbetalingsdato,
    }: Pick<ManedUtbetaling, "annenMottaker" | "forfallsdato" | "utbetalingsdato">,
    filters: FilterPredicate
) => {
    const matchMottaker = !filters.mottaker
        ? true
        : (filters.mottaker === "annenMottaker" && annenMottaker) ||
          (filters.mottaker === "minKonto" && !annenMottaker);

    const utbetalingDate = stringToDateWithoutTimezone(utbetalingsdato ?? forfallsdato);

    if (!utbetalingDate) return matchMottaker;

    const matchFra = !filters.fraDato ? true : !isBefore(utbetalingDate, filters.fraDato);
    const matchTil = !filters.tilDato ? true : !isAfter(utbetalingDate, filters.tilDato);

    return matchMottaker && matchTil && matchFra;
};
