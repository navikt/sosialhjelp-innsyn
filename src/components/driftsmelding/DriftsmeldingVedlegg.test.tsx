import { render } from "@testing-library/react";

import DriftsmeldingVedlegg from "./DriftsmeldingVedlegg";
import { useFileUploadError } from "./lib/useFileUploadError";

jest.mock("./lib/useFileUploadError", () => ({
    useFileUploadError: jest.fn(),
}));

describe("DriftsmeldingVedlegg", () => {
    it("renders null when there is no file upload error", () => {
        (useFileUploadError as jest.Mock).mockReturnValue(null);
        expect(render(<DriftsmeldingVedlegg />).asFragment()).toMatchSnapshot();
    });

    it("renders DriftsmeldingAlert when there is a file upload error", () => {
        (useFileUploadError as jest.Mock).mockReturnValue("driftsmelding.kanIkkeSendeVedlegg");
        expect(render(<DriftsmeldingVedlegg />).asFragment()).toMatchSnapshot();
    });

    it("renders DriftsmeldingAlert if sÃ¸knadinfo.isBroken is true", () => {
        (useFileUploadError as jest.Mock).mockReturnValue("driftsmelding.vedlegg.vedleggMangler");
        expect(render(<DriftsmeldingVedlegg className="custom-class" />).asFragment()).toMatchSnapshot();
    });
});
