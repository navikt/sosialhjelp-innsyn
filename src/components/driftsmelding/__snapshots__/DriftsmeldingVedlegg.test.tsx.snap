// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DriftsmeldingVedlegg renders DriftsmeldingAlert if søknadinfo.isBroken is true 1`] = `
<DocumentFragment>
  <div
    class="font-bold custom-class navds-alert navds-alert--error navds-alert--medium navds-alert--inline"
  >
    <svg
      aria-labelledby="title-r1"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r1"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M7.742 2.47a.75.75 0 0 1 .53-.22h7.456a.75.75 0 0 1 .53.22l5.272 5.272c.141.14.22.331.22.53v7.456a.75.75 0 0 1-.22.53l-5.272 5.272a.75.75 0 0 1-.53.22H8.272a.75.75 0 0 1-.53-.22L2.47 16.258a.75.75 0 0 1-.22-.53V8.272a.75.75 0 0 1 .22-.53zm1.288 5.5a.75.75 0 0 0-1.06 1.06L10.94 12l-2.97 2.97a.75.75 0 1 0 1.06 1.06L12 13.06l2.97 2.97a.75.75 0 1 0 1.06-1.06L13.06 12l2.97-2.97a.75.75 0 0 0-1.06-1.06L12 10.94z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-alert__wrapper--maxwidth navds-body-long navds-body-long--medium"
    >
      Du kan dessverre ikke sende inn dokumentasjon til denne søknaden.
    </div>
  </div>
</DocumentFragment>
`;

exports[`DriftsmeldingVedlegg renders DriftsmeldingAlert if the søknadinfo.isBroken flag is set 1`] = `
<DocumentFragment>
  <div
    class="font-bold custom-class navds-alert navds-alert--error navds-alert--medium navds-alert--inline"
  >
    <svg
      aria-labelledby="title-r1"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r1"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M7.742 2.47a.75.75 0 0 1 .53-.22h7.456a.75.75 0 0 1 .53.22l5.272 5.272c.141.14.22.331.22.53v7.456a.75.75 0 0 1-.22.53l-5.272 5.272a.75.75 0 0 1-.53.22H8.272a.75.75 0 0 1-.53-.22L2.47 16.258a.75.75 0 0 1-.22-.53V8.272a.75.75 0 0 1 .22-.53zm1.288 5.5a.75.75 0 0 0-1.06 1.06L10.94 12l-2.97 2.97a.75.75 0 1 0 1.06 1.06L12 13.06l2.97 2.97a.75.75 0 1 0 1.06-1.06L13.06 12l2.97-2.97a.75.75 0 0 0-1.06-1.06L12 10.94z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-alert__wrapper--maxwidth navds-body-long navds-body-long--medium"
    >
      Du kan dessverre ikke sende inn dokumentasjon til denne søknaden.
    </div>
  </div>
</DocumentFragment>
`;

exports[`DriftsmeldingVedlegg renders DriftsmeldingAlert when there is a file upload error 1`] = `
<DocumentFragment>
  <div
    class="font-bold navds-alert navds-alert--error navds-alert--medium navds-alert--inline"
  >
    <svg
      aria-labelledby="title-r0"
      class="navds-alert__icon"
      fill="none"
      focusable="false"
      height="1em"
      role="img"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="title-r0"
      >
        Feil
      </title>
      <path
        clip-rule="evenodd"
        d="M7.742 2.47a.75.75 0 0 1 .53-.22h7.456a.75.75 0 0 1 .53.22l5.272 5.272c.141.14.22.331.22.53v7.456a.75.75 0 0 1-.22.53l-5.272 5.272a.75.75 0 0 1-.53.22H8.272a.75.75 0 0 1-.53-.22L2.47 16.258a.75.75 0 0 1-.22-.53V8.272a.75.75 0 0 1 .22-.53zm1.288 5.5a.75.75 0 0 0-1.06 1.06L10.94 12l-2.97 2.97a.75.75 0 1 0 1.06 1.06L12 13.06l2.97 2.97a.75.75 0 1 0 1.06-1.06L13.06 12l2.97-2.97a.75.75 0 0 0-1.06-1.06L12 10.94z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="navds-alert__wrapper navds-alert__wrapper--maxwidth navds-body-long navds-body-long--medium"
    >
      Vi har dessverre problemer med systemet, og ber deg levere dokumentasjon på ditt Nav kontor
    </div>
  </div>
</DocumentFragment>
`;

exports[`DriftsmeldingVedlegg renders null when there is no file upload error 1`] = `<DocumentFragment />`;
